.buildkite/pipeline.yml

```diff
@@ -36,3 +36,15 @@ steps:
       echo "--- :rubocop: Run Rubocop"
       bundle exec rubocop
     plugins: *common_plugins
+
+  #################
+  # Push to RubyGems
+  #################
+  - label: ":rubygems: Publish to RubyGems"
+    key: "gem-push"
+    if: build.tag != null
+    depends_on:
+     - test
+     - rubocop
+    command: .buildkite/gem-push.sh
+    plugins: *common_plugins
```

Comment:
You'll need to have the docker plugin be told to forward the `RUBYGEMS_API_KEY` environment variable to its container too, otherwise when the `gem push` command will be run from within the docker container it will not know it.

You can do that like this when providing the plugin definition to `plugins:` like this
```yaml
    - docker#v3.8.0:
        image: &ruby_version "ruby:2.7.4"
        propagate-environment: true
        environment:
          - "RUBYGEMS_API_KEY"
```

There's a question weather it's ok to declare that `environment` to be propagated… as part of the `&common_plugin` YAML anchor — which means that we'd propagate this env var in all steps that use `plugins: *common_plugins`, including the ones that might not need that env var — or to create separate YAML anchors depending on which env vars each step that are using the docker plugin need to expose (which makes the configuration and the YAML a bit more cumbersome, so it is really worth it?)

Reply:
I wasn't aware of this but it makes sense, thanks.
I wonder if I could somehow reuse the `common_plugins` container with the right image and somehow just "override" the environment (I quickly searched in the Buildkite docs and couldn't find it).

So for now I just declared the container again only for that step, propagating the environment: 4ee5a0760b080b82f94cd3e741d7bde836de65e9

Reply:
Yeah it's not really possible sadly, at least not directly in YAML (we'd have to do some custom pre-processing of the YAML pipeline before piping it to `buildkite-agent pipeline upload` if we wanted to support that — see paaHJt-50A-p2).

What you could do though and that I'd recommend (mostly so that we keep the constants and list of plugin versions used throughout the pipeline all grouped at the top for easier future maintenance) is to [create YAML anchors for the versions and images (and any other entries or strings you're repeating)](https://mobile.blog/2021/09/09/clean-config-files-with-yaml-anchors/), even if those anchors don't encapsulate the whole YAML node describing the whole plugin.

Something like this, for example:
```yaml
common_params:
  - &docker_plugin
    docker#v5.7.0
  - &docker_ruby_image
    "public.ecr.aws/docker/library/ruby:3.2.2"
  - &common_plugins
    - *docker_plugin :
        image: *docker_ruby_image

steps:
  - …
  - label: "🧹 Lint (Rubocop)"
    key: rubocop
    command: …
    plugins: *common_plugins
  - …
  - label: ":rubygems: Publish to RubyGems"
    key: "gem-push"
    if: build.tag != null
    depends_on:
     - test
     - rubocop
    command: .buildkite/gem-push.sh
    plugins:
      - *docker_plugin :
          image: *docker_ruby_image
          propagate-environment: true
          environment:
            - "RUBYGEMS_API_KEY"
```
The only gotcha to know about is that if you want to use a string anchor as a dictionary key (like `*docker_plugin` in the example above to use the `docker_plugin` YAML anchor pointing to the `docker#v5.7.0` string), you need to be sure to have a space between the anchor name and the `:` — i.e. `*docker_plugin :` not just `*docker_plugin:`.

Reply:
> to [create YAML anchors for the versions and images (and any other entries or strings you're repeating)](https://mobile.blog/2021/09/09/clean-config-files-with-yaml-anchors/)

Thanks for the suggestions -- nice article, by the way 🙂

Reply:
Updated here: 337c46337acca77527f6ca08224f2fa3b0a8b13c

Reply:
I guess you meant 43debc1af5de8ccbc25abed59234f98f4a25c2dd 😛